### PR TITLE
Introduce IterableOnceOps as common abstraction for Iterable & Iterator

### LINKED ISSUE #465: Enforce consistency between the available operations of IterableOnce and Iterable
Currently these collection types don’t have a common super type.

@lrytz suggested that we write some tests that use reflection to check that the operations and type signatures are consistent between them.

### PR DESCRIPTION
**Pull request by szeiger**
Fixes #465

### CODE REVIEW

**Comment by szeiger** [comment]
Looks like an actual unsoundness in `SortedMultiSet` slipped through the cracks of the compiler(s) so far but gets flagged now that `zipWithIndex` was moved to a supertrait.

File: collections-contrib/src/main/scala/strawman/collection/immutable/MultiSet.scala [hunk 1]
```diff
@@ -10,10 +10,38 @@ import strawman.collection.decorators.ImmutableMapDecorator
...
   */
-class MultiSet[A] private (elems: Map[A, Int])
+trait MultiSet[A]
   extends collection.MultiSet[A]
     with Iterable[A]
-    with collection.MultiSetOps[A, MultiSet, MultiSet[A]] {
+    with MultiSetOps[A, MultiSet, MultiSet[A]] {
+
+  override def iterableFactory: IterableFactory[MultiSet] = MultiSet
+}
```
Comment (julienrf) [inline collections-contrib/src/main/scala/strawman/collection/immutable/MultiSet.scala]: Can you elaborate on the problem that is solved by introducing the separation `MultiSet`/`MultiSetImpl`?

Comment (julienrf) [reply collections-contrib/src/main/scala/strawman/collection/immutable/MultiSet.scala]: Oh maybe that’s for making `SortedMultiSet` a subtype of `MultiSet`?

File: collections/src/main/scala/strawman/collection/IterableOnce.scala [hunk 2]
```diff
@@ -135,3 +138,831 @@ object IterableOnce {
   @`inline` implicit def iterableOnceExtensionMethods[A](it: IterableOnce[A]): IterableOnceExtensionMethods[A] =
     new IterableOnceExtensionMethods[A](it)
 }
+
+/** This implementation trait can be mixed into an `IterableOnce` to get the basic methods that are shared between
+  * `Iterator` and `Iterable`. The `IterableOnce` must support multiple calls to `iterator()` but may or may not
+  * return the same `Iterator` every time.
+  */
```
Comment (julienrf) [inline collections/src/main/scala/strawman/collection/IterableOnce.scala]: You should define the documentation variables that are not inherited but used in the operations’ scaladoc.

File: collections-contrib/src/main/scala/strawman/collection/immutable/MultiSet.scala (hunk 1 above)
Comment (szeiger) [reply collections-contrib/src/main/scala/strawman/collection/immutable/MultiSet.scala]: Yes, you want the usual diamond pattern with all of:

- mutable.SortedMultiSet <: collection.SortedMultiSet
- immutable.SortedMultiSet <: collection.SortedMultiSet
- mutable.MultiSet <: collection.MultiSet
- immutable.MultiSet <: collection.MultiSet

- collection.SortedMultiSet <: collection.MultiSet
- mutable.SortedMultiSet <: mutable.MultiSet
- immutable.SortedMultiSet <: immutable.MultiSet

This requires that `MultiSet` is a trait.

**Pull request merged**